sql/handler.cc

```diff
@@ -6456,6 +6470,91 @@ void handler::set_lock_type(enum thr_lock_type lock)
   table->reginfo.lock_type= lock;
 }
 
+int handler::ha_check_overlaps(const uchar *old_data, const uchar* new_data)
+{
+  DBUG_ASSERT(new_data);
+  if (!table_share->period.unique_keys)
+    return 0;
```

Comment:
Please make sure that historical rows were not checked against OVERLAPS:

```sql
create or replace table t1 (id int, s date, e date,
                          period for p(s,e),
                          primary key(id, p without overlaps)) with system versioning;
insert into t1 values (1, '2002-01-01', '2002-02-01'),
                     (1, '2002-02-01', '2002-03-01');
delete from t1 where s = '2002-02-01';
```

You can check for `table->versioned() && !table->vers_row_end()->is_max()`.

Reply:
since it is possible to modify history data, we can't do this. Even historical rows should meet the constraints. And by design we could call `ha_check_overlaps` from `sql_*` side

But i was also keeping this optimization in mind. The solution i was thinking of is to add a possibility to enable constraint bypass (like `disable_overlaps_check()` and `enable_overlaps_check()`. There are also some other places where the check could be safely bypassed, for example the inserts made FOR PORTION OF time.

Reply:
BTW maybe we should consider making SYSTEM_TIME essentially WITHOUT OVERLAPS

Reply:
// Indeed, historical data ought not satisfy any actual constraints